### PR TITLE
feat(app): display row object

### PR DESCRIPTION
**Pull request by Kouta-fd**
- add single row display

スクリーンショット 2023-09-21 午後5 11 36

### CODE REVIEW

File: packages/app/src/pages/endpoints/[endpointId]/_/content/types/table/index.tsx [hunk 1]
```diff
@@ -174,3 +212,143 @@ const Operations: React.FC<OperationsProps> = ({
     </>
   );
 };
+
+const RowObject: React.FC<
+  BaseProps & { rowObject: Data; columns: TableColumn[]; close: () => void }
+> = ({ on, rowObject, columns, close }) => {
+  return (
+    <div className="flex flex-col h-full w-full">
```
Comment (nonoakij) [inline packages/app/src/pages/endpoints/[endpointId]/_/content/types/table/index.tsx:220]: 子どもにつけてるpadding はこの要素に集約できそう見えました！

File: packages/app/src/pages/endpoints/[endpointId]/_/content/types/table/index.tsx [hunk 2]
```diff
@@ -174,3 +212,143 @@ const Operations: React.FC<OperationsProps> = ({
...
+const Container: React.FC<
+  BaseProps & {
+    columns: TableColumn[];
+    rowObject: Data;
+    index: number;
+    objectKey: string;
+  }
+> = ({ on, columns, rowObject, index, objectKey }) => {
+  const schema = (data: TableColumn[] | Data, objectKey: string) => {
+    if (Array.isArray(data)) {
```
Comment (nonoakij) [inline packages/app/src/pages/endpoints/[endpointId]/_/content/types/table/index.tsx:259]: data は TableColumn[] だから isArray出ない時はないような気がします

File: packages/app/src/pages/endpoints/[endpointId]/_/content/types/table/index.tsx [hunk 3]
```diff
@@ -174,3 +212,143 @@ const Operations: React.FC<OperationsProps> = ({
...
+    () => (
+      <Button
+        variant="text"
+        on={on}
+        Icon={isOpened ? ChevronDownIcon : ChevronRightIcon}
+        onClick={handleArrowClick}
+      />
+    ),
+    [on, isOpened, handleArrowClick]
+  );
```
Comment (nonoakij) [inline packages/app/src/pages/endpoints/[endpointId]/_/content/types/table/index.tsx:281]: メモしてる理由が気になりました！
いらないようにも見えます！

File: packages/app/src/pages/endpoints/[endpointId]/_/content/types/table/index.tsx [hunk 4]
```diff
@@ -174,3 +212,143 @@ const Operations: React.FC<OperationsProps> = ({
...
+          const value = 'FALSE';
+          return value;
+        }
+      default:
+        return data;
+    }
+  };
+
+  return (
+    <div key={index}>
```
Comment (nonoakij) [inline packages/app/src/pages/endpoints/[endpointId]/_/content/types/table/index.tsx:305]: key いらなそうですが、どうでしょう？

File: packages/app/src/pages/endpoints/[endpointId]/_/content/types/table/index.tsx [hunk 5]
```diff
@@ -104,6 +112,18 @@ const ContentTable: React.FC<Props> = ({
     [document, base, sorts]
   );
 
+  const [rowObject, setRowObject] = useState<Data>();
```
Comment (h8570rg) [inline packages/app/src/pages/endpoints/[endpointId]/_/content/types/table/index.tsx:115]: It doesn't seem to be like ''specified(selected)" and "specified data type".
If its type is Data, I think selectedRow,  selectedRowData or selectedData is better.

File: packages/app/src/pages/endpoints/[endpointId]/_/content/types/table/index.tsx [hunk 6]
```diff
@@ -104,6 +112,18 @@ const ContentTable: React.FC<Props> = ({
...
+  const onRowClick = useCallback(
```
Comment (h8570rg) [inline packages/app/src/pages/endpoints/[endpointId]/_/content/types/table/index.tsx:116]: We often use not on but handle for click handler.

File: packages/app/src/pages/endpoints/[endpointId]/_/content/types/table/index.tsx [hunk 7]
```diff
@@ -104,6 +112,18 @@ const ContentTable: React.FC<Props> = ({
...
+    (data: Data) => {
+      drawer.open();
+      setRowObject(data);
+    },
+    [drawer]
+  );
+  const closeDrawer = () => {
+    drawer.close();
+  };
```
Comment (h8570rg) [inline packages/app/src/pages/endpoints/[endpointId]/_/content/types/table/index.tsx:125]: fyi It's possible to pass drawer.close to the component like
close={drawer.close}

File: packages/app/src/pages/endpoints/[endpointId]/_/content/types/table/index.tsx [hunk 8]
```diff
@@ -1,10 +1,16 @@
...
+import React, { useCallback, useMemo, useState } from 'react';
+import { Props as BaseProps } from '~/components';
 import Button, { Props as ButtonProps } from '~/components/button';
 import DotsCircleHorizontalIcon from '~/components/icon/dotsCircleHorizontal/outline';
 import Table, { Props as TableProps } from '~/components/table';
+import XIcon from '~/components/icon/xIcon/outline';
+import ChevronDownIcon from '~/components/icon/chevronDown/outline';
+import ChevronRightIcon from '~/components/icon/chevronRight/outline';
+import ClipboardCopyIcon from '~/components/icon/clipboardCopy/outline';
+import { Data } from '~/components/table';
```
Comment (h8570rg) [inline packages/app/src/pages/endpoints/[endpointId]/_/content/types/table/index.tsx:10]: Using Data, but not using Table? It's wierd.

File: packages/app/src/pages/endpoints/[endpointId]/_/content/types/table/index.tsx [hunk 9]
```diff
@@ -174,3 +205,143 @@ const Operations: React.FC<OperationsProps> = ({
...
   );
 };
+
+const RowObject: React.FC<
+  BaseProps & { rowObject: Data; columns: TableColumn[]; close: () => void }
+> = ({ on, rowObject, columns, close }) => {
+  return (
+    <div className="flex flex-col h-full w-full">
+      <div className="px-10 pt-10">
+        <div className="text-right">
```
Comment (h8570rg) [inline packages/app/src/pages/endpoints/[endpointId]/_/content/types/table/index.tsx:215]: Can you pass text-right to Button's className props?
It's not nice to have unnecessary divs.

File: packages/app/src/pages/endpoints/[endpointId]/_/content/types/table/index.tsx [hunk 10]
```diff
@@ -174,3 +205,143 @@ const Operations: React.FC<OperationsProps> = ({
...
+          <Button
+            variant="text"
+            on={COLOR_SYSTEM.SURFACE}
+            Icon={XIcon}
+            onClick={close}
+          />
+        </div>
+      </div>
+      <div className="px-10 pb-10 grow overflow-scroll">
+        <div>
```
Comment (h8570rg) [inline packages/app/src/pages/endpoints/[endpointId]/_/content/types/table/index.tsx:225]: Same as commended above, It seems to be unnecessary.

File: packages/app/src/pages/endpoints/[endpointId]/_/content/types/table/index.tsx [hunk 11]
```diff
@@ -174,3 +205,143 @@ const Operations: React.FC<OperationsProps> = ({
...
+          {Object.keys(rowObject).map((objectKey, index) => (
+            <div className="mt-10">
```
Comment (h8570rg) [inline packages/app/src/pages/endpoints/[endpointId]/_/content/types/table/index.tsx:227]: If the parent have space-y class, mt-10 and the div can be deleted.

File: packages/app/src/pages/endpoints/[endpointId]/_/content/types/table/index.tsx [hunk 12]
```diff
@@ -174,3 +205,143 @@ const Operations: React.FC<OperationsProps> = ({
...
+              />
+            </div>
+          ))}
+        </div>
+      </div>
+    </div>
+  );
+};
+
+const Container: React.FC<
```
Comment (h8570rg) [inline packages/app/src/pages/endpoints/[endpointId]/_/content/types/table/index.tsx:243]: What container?

File: packages/app/src/pages/endpoints/[endpointId]/_/content/types/table/index.tsx (hunk 5 above)
Comment (Kouta-fd) [reply packages/app/src/pages/endpoints/[endpointId]/_/content/types/table/index.tsx:115]: It means specified data, so I will change rowObject to selectedRowData.

File: packages/app/src/pages/endpoints/[endpointId]/_/content/types/table/index.tsx (hunk 6 above)
Comment (Kouta-fd) [reply packages/app/src/pages/endpoints/[endpointId]/_/content/types/table/index.tsx:116]: I will use `handle`  for a click handler.

File: packages/app/src/pages/endpoints/[endpointId]/_/content/types/table/index.tsx (hunk 7 above)
Comment (Kouta-fd) [reply packages/app/src/pages/endpoints/[endpointId]/_/content/types/table/index.tsx:125]: This function is not necessary so I will adopt your suggestion.

File: packages/app/src/pages/endpoints/[endpointId]/_/content/types/table/index.tsx (hunk 1 above)
Comment (Kouta-fd) [reply packages/app/src/pages/endpoints/[endpointId]/_/content/types/table/index.tsx:213]: Style properties that can be grouped together will be grouped together.

File: packages/app/src/pages/endpoints/[endpointId]/_/content/types/table/index.tsx (hunk 10 above)
Comment (Kouta-fd) [reply packages/app/src/pages/endpoints/[endpointId]/_/content/types/table/index.tsx:225]: I will avoid unnecessary divs.

File: packages/app/src/pages/endpoints/[endpointId]/_/content/types/table/index.tsx (hunk 9 above)
Comment (Kouta-fd) [reply packages/app/src/pages/endpoints/[endpointId]/_/content/types/table/index.tsx:215]: I will  pass text-right to Button's className props.

File: packages/app/src/pages/endpoints/[endpointId]/_/content/types/table/index.tsx (hunk 12 above)
Comment (Kouta-fd) [reply packages/app/src/pages/endpoints/[endpointId]/_/content/types/table/index.tsx:243]: I intended to "Accordion" so rename it.

File: packages/app/src/pages/endpoints/[endpointId]/_/content/types/table/index.tsx (hunk 11 above)
Comment (Kouta-fd) [reply packages/app/src/pages/endpoints/[endpointId]/_/content/types/table/index.tsx:227]: I will adopt your suggesion.

File: packages/app/src/pages/endpoints/[endpointId]/_/content/types/table/index.tsx (hunk 4 above)
Comment (Kouta-fd) [reply packages/app/src/pages/endpoints/[endpointId]/_/content/types/table/index.tsx:305]: I will deleted key attributes.
I am thinking of adding a key attirbute to `RowObeject` and `Container` .

File: packages/app/src/pages/endpoints/[endpointId]/_/content/types/table/index.tsx [hunk 13]
```diff
@@ -174,3 +205,143 @@ const Operations: React.FC<OperationsProps> = ({
...
+        <div className="ml-5 pl-4 border-l border-thm-on-surface-slight">
+          {typeof rowObject[objectKey] === 'object' ? (
+            Object.keys(rowObject[objectKey]).map((childObjectKey, index) => (
+              <Container
+                on={on}
+                columns={schema(columns, objectKey)}
+                rowObject={rowObject[objectKey]}
+                objectKey={childObjectKey}
+                index={index}
+              />
```
Comment (Kouta-fd) [inline packages/app/src/pages/endpoints/[endpointId]/_/content/types/table/index.tsx:327]: Passing objects to columns as props to this `Container`.

File: packages/app/src/pages/endpoints/[endpointId]/_/content/types/table/index.tsx (hunk 2 above)
Comment (Kouta-fd) [reply packages/app/src/pages/endpoints/[endpointId]/_/content/types/table/index.tsx:252]: data can be also an object.

Comment (Kouta-fd) [reply packages/app/src/pages/endpoints/[endpointId]/_/content/types/table/index.tsx:252]: Please check the comment on line 319.

File: packages/app/src/pages/endpoints/[endpointId]/_/content/types/table/index.tsx (hunk 3 above)
Comment (Kouta-fd) [reply packages/app/src/pages/endpoints/[endpointId]/_/content/types/table/index.tsx:281]: > Why are you memorizing this component?

The Button component expects to  re-run only when "on", "isOpened", and "handleArrowClick" are updated.

File: packages/app/src/pages/endpoints/[endpointId]/_/content/types/table/index.tsx (hunk 2 above)
Comment (nonoakij) [reply packages/app/src/pages/endpoints/[endpointId]/_/content/types/table/index.tsx:252]: It appears to me that in the actual use case, the Args of the schema function consists only of TableColumn[].

File: packages/app/src/pages/endpoints/[endpointId]/_/content/types/table/index.tsx [hunk 14]
```diff
@@ -27,6 +33,11 @@ type Props = {
     React.Dispatch<React.SetStateAction<Record<string, Sort>>>
   ];
 };
+
+type Value = any;
+type Key = string;
+type ObjectData = Record<Key, Value>;
```
Comment (h8570rg) [inline packages/app/src/pages/endpoints/[endpointId]/_/content/types/table/index.tsx:39]: What value, key and Object ?

Comment (h8570rg) [reply packages/app/src/pages/endpoints/[endpointId]/_/content/types/table/index.tsx:39]: You might be able to resolve Data type and naming issues if you change the placement of the showing row data function from this component to the Table component.

Comment (Kouta-fd) [reply packages/app/src/pages/endpoints/[endpointId]/_/content/types/table/index.tsx:39]: I will change the placement of the showing row data function from this component to the Table component in order to solve the issues.

File: packages/app/src/pages/endpoints/[endpointId]/_/content/types/table/index.tsx (hunk 2 above)
Comment (Kouta-fd) [reply packages/app/src/pages/endpoints/[endpointId]/_/content/types/table/index.tsx]: I will change to pass "schema" to Accordion's schema props.

File: packages/app/src/components/table/index.tsx [hunk 15]
```diff
@@ -237,3 +260,127 @@ const Td: React.FC<BaseProps & { isSticky?: boolean }> = ({
...
+const RowObject: React.FC<
+  BaseProps & {
+    rowObject: Data;
+    columns: TableColumn[];
+    close: () => void;
+  }
+> = ({ on, rowObject, columns, close }) => {
+  return (
+    <div className="px-10 py-10 flex flex-col h-full w-full">
+      <div>
```
Comment (h8570rg) [inline packages/app/src/components/table/index.tsx:273]:
```suggestion
```

needless?

File: packages/app/src/components/table/index.tsx [hunk 16]
```diff
@@ -28,80 +34,96 @@ const Table: React.FC<Props> = ({
...
+                </React.Fragment>
+              ))}
+            </tbody>
+          </table>
+        </div>
       </div>
-    </div>
+      {selectedRowData && (
+        <Drawer {...drawer.bind}>
+          <RowObject
```
[the remark this answers is not in the document]
Comment (h8570rg) [reply packages/app/src/components/table/index.tsx:118]: And this component doesn't care about if the data is coming from row or not, so some word to describe the data type. (but `Data` is not proper, hmm)

[the remark this answers is not in the document]
Comment (h8570rg) [reply packages/app/src/components/table/index.tsx:118]: however, RowData may be proper. this app seems to treat column and row separately. Row and Column is clearly different. (sorry!)

File: packages/app/src/components/table/index.tsx [hunk 17]
```diff
@@ -237,3 +260,127 @@ const Td: React.FC<BaseProps & { isSticky?: boolean }> = ({
...
+            selectedRowData={rowObject}
+            objectKey={objectKey}
+          />
+        ))}
+      </div>
+    </div>
+  );
+};
+
+const Accordion: React.FC<
```
Comment (h8570rg) [inline packages/app/src/components/table/index.tsx:297]: Can this component be used commonly? If so, move it /components/Accordion.

File: packages/app/src/components/table/index.tsx [hunk 18]
```diff
@@ -237,3 +260,127 @@ const Td: React.FC<BaseProps & { isSticky?: boolean }> = ({
...
+    schema: TableColumn['schema'] | undefined;
+    selectedRowData: Data;
+    objectKey: string;
+  }
+> = ({ on, schema, selectedRowData, objectKey }) => {
+  const [isOpened, setIsOpened] = useState<boolean>(true);
+  const handleArrowClick = useCallback<ButtonProps['onClick']>(() => {
+    setIsOpened(!isOpened);
+  }, [isOpened]);
+  const arrowIcon = useMemo<JSX.Element>(
```
Comment (h8570rg) [inline packages/app/src/components/table/index.tsx:308]: No need to memorize. write it inline

Comment (Kouta-fd) [reply packages/app/src/components/table/index.tsx:308]: I will write it inline.

File: packages/app/src/components/table/index.tsx (hunk 17 above)
Comment (Kouta-fd) [reply packages/app/src/components/table/index.tsx:297]: This component can be used commonly so move it the right place

File: packages/app/src/components/table/index.tsx (hunk 16 above)
[the remark this answers is not in the document]
Comment (Kouta-fd) [reply packages/app/src/components/table/index.tsx:118]: This component is for Data so RowData may be better.

File: packages/app/src/components/table/index.tsx (hunk 15 above)
Comment (Kouta-fd) [reply packages/app/src/components/table/index.tsx:273]: needless!

File: packages/app/src/components/Accordion/index.tsx [hunk 19]
```diff
@@ -0,0 +1,91 @@
...
+          return value;
+        }
+      default:
+        return data;
+    }
+  };
+
+  return (
+    <div>
+      <div className="flex-none inline-flex items-center gap-1 whitespace-nowrap">
```
Comment (nonoakij) [inline packages/app/src/components/Accordion/index.tsx:43]:
```suggestion
      <div className="inline-flex items-center gap-1 whitespace-nowrap">
```

**Comment by Kouta-fd** [comment]
I will move accordion component `/components/table/`

File: packages/app/src/components/table/index.tsx (hunk 17 above)
Comment (Kouta-fd) [reply packages/app/src/components/table/index.tsx]: This accordion only accept object, so can not say it is universal. I will move accordion component `/components/table/`

File: packages/app/src/components/table/index.tsx [hunk 20]
```diff
@@ -237,3 +260,118 @@ const Td: React.FC<BaseProps & { isSticky?: boolean }> = ({
...
+      case 'number':
+        return data.toLocaleString();
+      case 'boolean':
+        if (data === true) {
+          const value = 'TRUE';
+          return value;
+        } else {
+          const value = 'FALSE';
+          return value;
+        }
```
Comment (h8570rg) [inline packages/app/src/components/table/index.tsx:323]: That could be

```suggestion
   return String(value).toUpperCase()
```

**Review by nonoakij — APPROVED** [review]

**Review by h8570rg — APPROVED** [review]

**Pull request merged**